### DISCUSSION TITLE
Add "url" support for OpenAI embedder

### DISCUSSION THREAD

**Discussion by DavertMik** [meilisearch/product#775] [Experimental features]
It's a known thing that lots of AI providers use OpenAI API as their standard.

It would be nice if we could use OpenAI API but specifying a host other than `api.openai.com`

This would allow to use Microsoft Azure OpenAI and other providers as DeepInfra  (which uses OpenAI API as well https://deepinfra.com/BAAI/bge-large-en-v1.5/api ) 

So I would suggest to extend configuration for OpenAI so we could use other vendors but the same API

**Comment by dureuill** [comment]
Hello :wave:

Thank you for the suggestion. It so happens that the upcoming v1.10 is adding the `url` parameter as an optional parameter :tada:
That said, using this new parameter for the use cases you describe is **not** recommended :warning: 

> This would allow to use Microsoft Azure OpenAI 

Microsoft Azure OpenAI uses non-standard headers for authentication.

> other providers as DeepInfra

These providers are not using the OpenAI API, they're using an OpenAI**-like** API. Some examples of the differences:

1. the model fields take a different value (`"BAAI/bge-large-en-v1.5"`)
2. they don't support inputs in the forms of tokens tokenized with OpenAI's tokenizer
3. the token limit is not the same as for OpenAI limit (8196 vs 512)

The `openAi` embedder of Meilisearch is specialized to communicate with a genuine OpenAI endpoint. The `url` parameter was mainly added to serve the use case of people who proxy requests to OpenAI.

Instead, you should use the `rest` embedder source. Meilisearch v1.10 makes this more convenient:

1. `rest` embedders support custom headers, allowing Microsoft Azure OpenAI
2. the configuration of `rest` embedders is much simpler, with the newly introduced `"request"` and `"response"` parameters. According to their request/response example, the DeepInfra embedder would have the following configuration:

```jsonc
"default": {
  "source": "rest",
  "url": "https://api.deepinfra.com/v1/openai/embeddings",
  "apiKey": "<YOUR_DEEPINFRA_TOKEN>",
  "request": {
    "input": "{{text}}",
    "model": "BAAI/bge-large-en-v1.5",
    "encoding_format": "float"
  },
  "response": {
    "data": [
      {
        "embedding": "{{embedding}}"
      }
    ]
  }
}
```

or, to optimize throughput, as it seems supported by DeepInfra:

```jsonc
"default": {
  "source": "rest",
  "url": "https://api.deepinfra.com/v1/openai/embeddings",
  "apiKey": "<YOUR_DEEPINFRA_TOKEN>",
  "request": {
    "input": ["{{text}}", "{{..}}"],
    "model": "BAAI/bge-large-en-v1.5",
    "encoding_format": "float"
  },
  "response": {
    "data": [
      {
        "embedding": "{{embedding}}"
      },
      "{{..}}"
    ]
  }
}
```

 [More information about the newly added `url` parameter](https://meilisearch.notion.site/v1-10-AI-search-changes-737c9d7d010d4dd685582bf5dab579e2#bcca142c854e4c77ad041ab41055f42b).

[More information about the new REST embedder API](https://meilisearch.notion.site/v1-10-AI-search-changes-737c9d7d010d4dd685582bf5dab579e2#8de842673ffa4a139210094a89c1ec3e)

**Reply by DavertMik** [reply]
Thanks for the detailed explanation on this issue!
Really appreciated. I was reviewing RC release notes and didn't find mention of `url` param in there so I started this discussion.
However, your reply is more valuable than that, I will consider using REST embedder once the 1.10 is out!

Thanks so much!

**Reply by dureuill** [reply]
Happy to be of service :-)

I will close the discussion, feel free to reopen as needed :sunny:
